python/tank/bootstrap/configuration_writer.py

```diff
@@ -119,76 +163,86 @@ def get_descriptor_metadata_file(self):
-            # sometimes, execution and rollback is so quick that several backup folders
-            # are created in a single second. In that case, append a suffix
-            core_backup_path = os.path.join(core_backup_root, timestamp)
-            counter = 0
-            while os.path.exists(core_backup_path):
-                # that backup path already exists. Try another one
-                counter += 1
-                core_backup_path = os.path.join(core_backup_root, "%s.%d" % (timestamp, counter))
-
-            log.debug("Moving core %s -> %s" % (core_payload, core_backup_path))
-            os.rename(core_payload, core_backup_path)
-            log.debug("Backup complete.")
-            core_backup_path = core_backup_path
-
-        return (config_backup_path, core_backup_path)
+        # The rename guard's role is to keep track of every rename operation that has happened
+        # in a given scope and undo all the renames if something went wrong. This feels a lot simpler
+        # than having a try/except block as we'd have to have different try/except blocks for different
+        # code sections. Its also a great way to avoid having to deal with variables that haven't been
+        # defined yet when dealing with the exceptions.
+        with RenameGuard(undo_on_error) as guard:
+            config_backup_path = None
+            core_backup_path = None
+
+            # get backup root location
+            config_path = self._path.current_os
+            configuration_payload = os.path.join(config_path, "config")
+
+            # timestamp for rollback backup folders
+            timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
+
+            if os.path.exists(configuration_payload):
+
+                config_backup_root = os.path.join(config_path, "install", "config.backup")
+
+                # make sure we have a backup folder present
+                # sometimes, execution and rollback is so quick that several backup folders
+                # are created in a single second. In that case, append a suffix
+                config_backup_path = os.path.join(config_backup_root, timestamp)
+                counter = 0
+                while os.path.exists(config_backup_path):
+                    # that backup path already exists. Try another one
+                    counter += 1
+                    config_backup_path = os.path.join(config_backup_root, "%s.%d" % (timestamp, counter))
+
+                # now that we have found a spot for our backup, make sure folder exists
+                # and then move the existing config *into* this folder.
+                filesystem.ensure_folder_exists(config_backup_path)
+
+                log.debug("Moving config %s -> %s" % (configuration_payload, config_backup_path))
+                backup_target_path = os.path.join(config_backup_path, os.path.basename(configuration_payload))
+                guard.rename(configuration_payload, backup_target_path)
```

Comment:
Gotcha. I actually didn't see the second `guard.rename()` call yesterday. Seems perfectly reasonable. Thanks for clarifying!